### ISSUE TITLE
Set up InkCut for usage with vinyl cutters.

### ISSUE THREAD

**Issue by MatthewCroughan** [DoESLiverpool/somebody-should#1208]
This looks like really promising software, much better than anything we have currently. 
https://github.com/inkcut/inkcut
#990

![Inkcut - vinyl cutting and plotting](https://user-images.githubusercontent.com/380158/34273634-29e60a08-e663-11e7-9977-5125eae738f7.gif)

**Comment by MatthewCroughan** [comment]
I've attempted this in Little Sandbox with a random still unknown vinyl cutter with Inkcut's generic cutter driver. It worked great, without adjustment, even cut out a piece on the first try by accident.

### Video Demonstration (Click to play)

[<img src="https://img.youtube.com/vi/3U3cz_ud8BE/maxresdefault.jpg" width="50%">](https://youtu.be/3U3cz_ud8BE)

### Cutter Output

[<img src="https://user-images.githubusercontent.com/26458780/66663340-72d3f180-ec42-11e9-8059-fba7bde87ad3.png" width="50%">]

**Comment by MatthewCroughan** [comment]
I've created a balena based application for the Roland vinyl cutter that I'm going to start documenting and upload to github soon once I've made it more generic, rather than hard coded. 

This means we can finally address #1127 and start deploying services on pi-like single board computers at the very least. This is a proof of concept.

Inkcut works perfectly with the Roland cutter. 

This is currently using Xorg, though I plan to use Wayland via Weston to create a generally better experience with less frame tearing, though that is a nitpick considering this runs well on a Pi at all.

![image](https://user-images.githubusercontent.com/26458780/67626011-a2d3e500-f83d-11e9-9975-c5dee9e5c707.png)

![image](https://user-images.githubusercontent.com/26458780/67625964-027dc080-f83d-11e9-8423-9833df5beaa6.png)

The image runs great on a raspberry pi 3 and provides:

- nautilus file manager
- xfce4-terminal
- inkscape
- inkcut
- chromium web browser

and most importantly access to the shared samba drive via an envvar $SAMBA_SERVER which can be set via the balena cli or the balena cloud web interface.

I'll document this in detail when I get back from the Astral ship on Monday.

At the current time the dockerfile looks something like 
```yml
FROM balenalib/%%BALENA_MACHINE_NAME%%-debian:buster-run

# Set Workdir
WORKDIR /usr/src/app

# Install XORG
RUN install_packages xserver-xorg \
  xserver-xorg-input-evdev \
  xinit \
  xfce4-session \
  xfce4-settings \
  xfce4-terminal \
  xfdesktop4 \
  libxfce4ui-utils \
  gtk2-engines-xfce \
  xfce4-pulseaudio-plugin \
  xfconf \
  xfwm4 \
  x11-xserver-utils \
  dbus-x11 \
  cifs-utils

# Install DE apps
RUN install_packages arc-theme \
  inkscape \
  nautilus \
  plank \
  chromium

# Install Inkcut
RUN install_packages gcc build-essential \
  python3-pip \
  python3-dev \
  python3-wheel \
  python3-setuptools \
  python3-pyqt5.qtsvg \
  python3-pyqt5 \
  libatlas-base-dev \
  libcups2-dev \
  && pip3 install inkcut --no-cache-dir

# Install papirus icons
RUN echo "deb [trusted=yes] http://ppa.launchpad.net/papirus/papirus/ubuntu bionic main" > /etc/apt/sources.list.d/papirus-ppa.list \
  && apt-get update -qq \
  && apt-get install papirus-icon-theme -y --no-install-recommends \
  && apt-get remove xfce4-panel wget thunar gcc build-essential -y \
  && apt-get clean \
  && rm -rf /tmp/* /var/tmp/*  \
  && rm -rf /var/lib/apt/lists/*

# Disable screen from turning it off
#RUN echo "#!/bin/bash" > /etc/X11/xinit/xserverrc \
#  && echo "" >> /etc/X11/xinit/xserverrc \
#  && echo 'exec /usr/bin/X -s 0 dpms' >> /etc/X11/xinit/xserverrc

COPY ./dockerbin/config/ /root/.config/
COPY ./dockerbin/plank/themes /root/.local/share/plank/themes
COPY ./dockerbin/inkcut.desktop /usr/share/applications/inkcut.desktop
COPY . ./

ENV UDEV=1

# Avoid requesting XFCE4 question on X start
ENV XFCE_PANEL_MIGRATE_DEFAULT=1

CMD ["bash", "start.sh"]
```

The copies in this file need to be reduced to nothing but environment variables such that other people can benefit.

**Comment by chris-does** [comment]
This looks great @MatthewCroughan :-)

Hope you have a great time at the Astral Ship, is it [this](https://astralship.org/voyages/VsgwrsCymunedolOct19.html) event you are attending?

It sounds very interesting!

If you are going there, and have the opportunity, may I ask if you could enquire if the Astral Ship has any links with Bangor University (in particulate their CS and Media faculties)?

Thank you! :-)

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by MatthewCroughan** [comment]
@chris-does I can do indeed, just getting up now and will ask whenever I arrive. You've got the right event.

**Comment by johnmckerrell** [comment]
This does sound great but currently the instructions on the wiki are incorrect then? Can you update them soon? https://github.com/DoESLiverpool/somebody-should/wiki/RolandVinylCutter We've had an enquiry from someone who wants to get started on the vinyl cutter but right now I don't know what to tell them about using it other than vaguely direct them here.

**Comment by MatthewCroughan** [comment]
@johnmckerrell Will get on this tomorrow.

**Comment by MatthewCroughan** [comment]
@johnmckerrell Will get on this tomorrow.

**Comment by MatthewCroughan** [comment]
@johnmckerrell I saw the person who was inquiring and helped them use the cutter in person. Documentation is secondary because everyone who has ever been to DoES primarily learns from people first and wiki second. I need help documenting, I can't do 100% of the work alone. So if anyone wants to volunteer to make notes with me on a night like Maker Night, that'd be great.

First we need to get everything working, secondly we can document. Everything here currently works, however there's an annoying quirk in Inkcut with this specific vinyl cutter.
https://github.com/inkcut/inkcut/issues/193

I want to try and use the cutter with its serial port rather than its parallel port, that way I can see if this is also an issue there. If it's not, then we can circumvent this issue. This is not an issue on the Little Sandbox vinyl cutter, for example, though it uses serial.

In order to test this I need a DB25 serial to USB. Unless I can get that, I can't move on testing that.

The solution for this problem is noted in the Inkscape issue. Upon launching Inkcut, the user needs to go to the Setup menu and change the language on the machine to anything else, and then back to HPGL, otherwise the controller on the vinyl cutter is not send the appropriate commands to get in into a usable state where it responds with "IN;"

**Comment by MatthewCroughan** [comment]
@goatchurchprime has made progress towards https://github.com/inkcut/inkcut/issues/45 as part of Hack The Space tonight.

**Comment by MatthewCroughan** [comment]
I've also asked @magman2112 to provide a DB25 since he suggested he may have one spare.

**Comment by MatthewCroughan** [comment]
@amcewen I can't get this done until I get a DB25 serial as Inkcut has issues me and Julian are looking into resolving but is a level of complexity I'd rather not bother with when the issue is solved by using Serial instead.

I gave this a hard go last night but never got any further in documenting. However, I did set up a new PC to do a better job than the Pi was doing and organised the cables massively using the new 3d printed cable clips, the area is hopefully now more effectively used as the monitor is sitting atop the PC case.  I also opened a new issue I found in Inkcut https://github.com/inkcut/inkcut/issues/225

I will document the usage of Inkcut once I have a DB25 Serial Cable.
I will then document the existence of this PC. 
I will continue developing previously discussed Balena image so that it can be used as a drop in replacement when the time comes, so we have a dockerfile that documents how we got to a working system.

The short term result will be a 32GB disk image of Ubuntu 18.04 we can backup and restore once we have verified that my claim of the serial cable solving all of our software woes is accurate, as I've tested on one of the aforementioned cutters in Little Sandbox.

**Comment by MatthewCroughan** [comment]
https://www.ebay.co.uk/itm/DB9-Male-to-DB25-Female-Serial-AT-Adapter/303132842448?hash=item46942019d0:g:uwUAAOSwpCFcvWn0

I've bought this, hopefully someone can either provide one sooner (Tomorrow) and a DB9 serial cable?

**Comment by MatthewCroughan** [comment]
@RussCoty Maybe you could help with this, you're the only person I know who might be old enough to have one of these connectors :scream:

**Comment by MatthewCroughan** [comment]
The one I ordered is the wrong gender on both ends, so I've ordered the correct one instead.

![image](https://user-images.githubusercontent.com/26458780/74790088-7e998980-52ae-11ea-8670-1c9ee3ea59e0.png)


![image](https://user-images.githubusercontent.com/26458780/74790132-ac7ece00-52ae-11ea-9422-38fda854eb5b.png)

**Comment by amcewen** [comment]
As we've had two enquiries about using the vinyl cutter this week, this evening I tried moving it back to the old computer to get it working again.

Sadly it's not being recognised by the computer at all, so it's still not working :-(

**Comment by MatthewCroughan** [comment]
@amcewen I'm sorry to hear that, check event-viewer to see what the response is? Have you tried it in Win10? Is it because Win7 update servers are down and it can't get the drivers for the vinylcutter automatically?

**Comment by MatthewCroughan** [comment]
My efforts with the serial cable were fruitless, as it turns out the CX24 Camm-1 Roland cutter uses RS232C (Crossover) serial which is very much unexpected. Serial is still preferable, but upon reading the Inkcut documentation it became clear that in order to properly utilise Parallel ports, one has to add a printer in CUPS (print server for linux) and set the driver type as "RAW". Which means raw output, essentially allowing us to send the data we want through the port uninhibited by quirks.

Once this is complete, Inkcut no longer has issues relating to starting and stopping jobs as it previously did, as CUPS is handling the port correctly for us. This means that in Inkcut ee select a printer rather than the raw device at `/dev/lp0`. Inkcut poorly manages raw access to the parallel port for reasons that probably have something to do with the nature and history of parallel, hence why the documentation suggests using CUPS. Though it is odd that it would work some of the time. 

All that is left to do is document usage of the machine and make a list of instructions of how to rebuild the machine and perhaps an ansible script to reprovision the machine from a base image of ubuntu 19.04.

**Comment by DoESsean** [comment]
Does it work then?

**Comment by MatthewCroughan** [comment]
@DoESsean Yes. Just need to show people how to use the new software, document that on the wiki etc.

It is however fairly straight forward if anyone wants to play with it without the help. Simply load an SVG into inkcut and print it. CTRL + P is the shortcut for printing.

I wouldn't close this issue until I have filled out the wiki though, which I assure you is the focus of maker night for me, tonight.

**Comment by MatthewCroughan** [comment]
Accidentally misclicked! 

What I was going to type was that in the end people won't really need to leave Inkscape as there's a plugin for all of this, so it will be just like printing a page on a regular printer.

**Comment by DoESsean** [comment]
Sweet. I might have to pop in tonight and get some stuff cut if I have chance.

**Comment by MatthewCroughan** [comment]
https://github.com/brandonedens/cups-fab/

This looks good too for future exploration, as it would allow us to have the machines be accessible by any machine running inkcut/inkscape over the network like a regular printer.

**Comment by MatthewCroughan** [comment]
https://github.com/DoESLiverpool/somebody-should/wiki/RolandVinylCutter/b8695751d2bebe3926f3e22d3f58cd3aa48e098e

I have updated details in the wiki, this is the specific revision.

**Comment by MatthewCroughan** [comment]
https://github.com/DoESLiverpool/somebody-should/wiki/RolandVinylCutter/a2db79c7b89593a768ce652bab67de43c48ae779

I have updated the formatting slightly, this is the specific revision

**Comment by MatthewCroughan** [comment]
Feel free to close this issue if you agree that the Wiki is now up to date @amcewen

**Comment by arthurrowland** [comment]
Wiki looks great @MatthewCroughan! ...but unfortunately today the PC won't boot.

Following a software crash (went to a command line and wouldn't respond) I rebooted it and the following error appeared
`isolinux.bin missing or corrupt`
and choosing 'ubuntu' via the boot menu doesn't get things any further.

How might I get it working again?

.

As a workaround I attempted to use my Windows 10 laptop. I successfully ran Inkcut 2.1.3 using Python 3.9.5, connected to the COM port and prepared the file for cutting, but it would immediately disconnect after sending the IN command and I couldn't get it to move or cut despite trying many things, including [saving the file and sending it via Hercules and using the QSerialPort setting](https://github.com/inkcut/inkcut/issues/217#issuecomment-715272844) - though I don't know if I was using Hercules correctly. The latter required PyQt5 and Twisted[windows_platform] Python libraries to stop throwing up errors. The log usually looked like this:

```
device | finish
-> COM11 IN;
device | disconnect
-- closed by request
```

The machine is not [officially supported](https://www.rolanddga.com/support/products/cutting/camm-1-cx-24-24-vinyl-cutter) beyond windows XP and the GX-24 64 bit driver didn't work for me, and neither did the inkcut generic driver.

**Comment by MatthewCroughan** [comment]
@amcewen The Vinylcutter has been set up similarly to #1580 which is well documented. The TL;DR is that the entire configuration for the machine is [here](https://github.com/DoESLiverpool/nixcfgs/blob/master/hosts/vinylcutter/configuration.nix#L64-L77). 

I will fill this issue out similarly to #1580 and close it today.

**Comment by ajlennon** [comment]
There are also a **host** of other issues with the Crealities I have had to sink huge amounts of time into.

Not least that the Ender 3D is unusable for the average person with the interface wheel disabled...

These are **not** improvements. Dig in and do work other people can maintain rather than making life hard for everybody else.

**Comment by MatthewCroughan** [comment]
@ajlennon Can you please make a separate issue for that and comment in there? This issue is about the Vinylcutter.

**Comment by MatthewCroughan** [comment]
> Wiki looks great @MatthewCroughan! ...but unfortunately today the PC won't boot.
> 
> Following a software crash (went to a command line and wouldn't respond) I rebooted it and the following error appeared
> `isolinux.bin missing or corrupt`
> and choosing 'ubuntu' via the boot menu doesn't get things any further.
> 
> How might I get it working again?
> 
> .
> 
> As a workaround I attempted to use my Windows 10 laptop. I successfully ran Inkcut 2.1.3 using Python 3.9.5, connected to the COM port and prepared the file for cutting, but it would immediately disconnect after sending the IN command and I couldn't get it to move or cut despite trying many things, including [saving the file and sending it via Hercules and using the QSerialPort setting](https://github.com/inkcut/inkcut/issues/217#issuecomment-715272844) - though I don't know if I was using Hercules correctly. The latter required PyQt5 and Twisted[windows_platform] Python libraries to stop throwing up errors. The log usually looked like this:
> 
> ```
> device | finish
> -> COM11 IN;
> device | disconnect
> -- closed by request
> ```
> 
> The machine is not [officially supported](https://www.rolanddga.com/support/products/cutting/camm-1-cx-24-24-vinyl-cutter) beyond windows XP and the GX-24 64 bit driver didn't work for me, and neither did the inkcut generic driver.

It is true that the Inkcut gui program is buggy with this particular vinyl cutter. I've been in IRC and talked about it. Inkscape itself has no issues sending directly to this cutter though. I will document the complete details of this tomorrow, as I recall this from anecdote.

The issue on the hardware seems to be that the SSD came loose from its sata port, which caused some data corruption. I've fixed the issue physically and installed NixOS on the machine.

**Comment by MatthewCroughan** [comment]
Last night, I got carried away making progress on building Windows images declarativelly using Nix on my stream here - https://www.twitch.tv/thenixzone/

A NixOS configuration for the Vinylcutter machine is publicly auditable and visible [here](https://github.com/DoESLiverpool/nixcfgs/blob/master/hosts/vinylcutter/configuration.nix#L71-L77)

This configuration will work on any hardware or machine. So we could for example flash a Raspberry Pi, old laptop, or any old computer with this software configuration and it'll be reproducible for a very long time if not forever.

Essentially all this is, is a list of software that should be on the computer and the way it should be configured to have the behavior we want, that anyone can add to on Github. This is known as declarative configuration. Another example of declarative configuration is [Ansible](https://www.ansible.com/).

It is opposed to someone running undocumented commands on the machine and having it be in an unknown state that is special and requires verbose textual documentation. 

In that scenario, it is impossible in computer science terms to get back to this state if you were to start from a fresh base image of the OS, as you would never run these undocumented commands again in the same way, in the same order, with the same network conditions, etc. In other words it would be non deterministic.

## Todo

- [x] Configure the samba share to be on the machine
- [ ] Document how to re-install this configuration anywhere, on any machine, should anything go wrong
  - [ ] Create a video demonstrating this installation process
  - [ ] Create a textual demonstration of this process

**Comment by MatthewCroughan** [comment]
It should also be mentioned that the machine should be usable in its current state by following the same wiki page. Though I haven't configured the Samba share yet, so that step will be missing.

**Comment by MatthewCroughan** [comment]
When I was last in (3 days ago), I had played around with the Vinyl Cutter to try and hook Inkcut up to CUPS so that it acts like a printer, in accordance with this https://github.com/DoESLiverpool/somebody-should/issues/1208#issuecomment-591860900.

I failed to do that, since there were problems in Inkcut which have are being looked into as a result of this effort:
https://github.com/inkcut/inkcut/pull/312
https://github.com/NixOS/nixpkgs/issues/130351

**Comment by MatthewCroughan** [comment]
I've created a reproducible config with the last 3 commits. The machine should not cease working due to user input now, since it'll be in a known state due to the Nix setup I've created, the user cannot modify the important bits that make the cutter work. I'd like to go even further in the future and make the machine wipe itself on reboot to a known good state, but this setup is more complex and will come later.

I consider this issue complete and will spin the checklist for other nice-to-haves out to new issues of their own.
